File: src/opt-hc.py
#!/usr/bin/env python3

import sys
import random
import networkx as nx
from argparse import ArgumentParser
from graph_config import GraphConfig
from edge_swap import random_swap

def optimize(G, conf, steps, log=None):
    def write_log_row(t, diam, aspl, ofile):
        print('{},{},{}'.format(t, diam, aspl), file=ofile)

    min_diam = nx.diameter(G)
    min_aspl = nx.average_shortest_path_length(G)
    if log:
        log_fp = open(log, 'w')
        print('t,diam,aspl', file=log_fp)
        write_log_row(0, min_diam, min_aspl, log_fp)
    for t in range(1, steps+1):
        H = G.copy()
        random_swap(H, conf)
        diam = nx.diameter(H)
        aspl = nx.average_shortest_path_length(H)
        if diam < min_diam or\
           diam == min_diam and aspl < min_aspl:
            G, min_diam, min_aspl = H, diam, aspl
        if log:
            write_log_row(t, min_diam, min_aspl, log_fp)

    if log:
        log_fp.close()
    return G

def get_options():
    parser = ArgumentParser()
    parser.add_argument('--size', type=int, required=True, nargs='+')
    parser.add_argument('-r', type=int, required=True, default=-1)
    parser.add_argument('-i', '--input', type=str, required=True)
    parser.add_argument('-o', '--output', type=str)
    parser.add_argument('--seed', type=int)
    parser.add_argument('--step', type=int, default=1000)
    parser.add_argument('--log', type=str)
    opts = parser.parse_args()
    return opts

def main():
    opts = get_options()
    random.seed(opts.seed)
    G = nx.read_edgelist(opts.input, nodetype=int)
    G = optimize(G, GraphConfig(opts.size, opts.r), opts.step, opts.log)

    os = open(opts.output, 'w') if opts.output else sys.stdout
    for e in nx.generate_edgelist(G, data=False):
        print(e, file=os)
    if opts.output:
        os.close()

if __name__ == '__main__':
    main()

File: src/opt-hc.cpp

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <boost/program_options.hpp>

#include "gg_util.hpp"
#include "graph_config.hpp"
#include "edge_swap.hpp"

using namespace std;
using namespace gg;
using namespace boost::program_options;

int main(int argc, char* argv[]) {
  string usage_doc =
    "usage: opt-hc.out [-h] --size SIZE [SIZE ...] -r R -i INPUT -o OUTPUT\n";
  usage_doc +=
    "                  [--seed SEED] --step STEP [--log LOG]";
  options_description desc("optional arguments");
  desc.add_options()
    ("help", "show this help message and exit")
    ("size", value<vector<int> >()->multitoken()->required(), "")
    (",r", value<int>()->default_value(-1), "")
    ("input", value<string>()->required(), "")
    ("output", value<string>()->required(), "")
    ("seed", value<unsigned long>(), "")
    ("step", value<int>()->default_value(1000), "")
    ("log", value<string>(), "");

  variables_map vm;
  try {
    store(command_line_parser(argc, argv)
          .options(desc)
          .style (command_line_style::default_style |
                  command_line_style::allow_guessing |
                  command_line_style::allow_long_disguise)
          .run(), vm);
    if(vm.count("help")) {
      cout << usage_doc << endl
           << desc << endl;
      return 0;
    }
    notify(vm);
  } catch(error& e) {
    cout << e.what() << endl
         << usage_doc << endl
         << desc << endl;
    return 1;
  }

  vector<int> size = vm["size"].as<vector<int> >();
  int r = vm["-r"].as<int>();
  GraphConfig conf(size, r);

  string input = vm["input"].as<string>();
  string output = vm["output"].as<string>();
  ofstream log_fs;
  if(vm.count("log"))
    log_fs.open(vm["log"].as<string>());
  if(vm.count("seed")) srand(vm["seed"].as<long int>());
  else srand(time(NULL));

  Graph G = read_edgelist(input);
  size_t n = num_vertices(G);

  pair<int, int> best_score = score(G);
  if(vm.count("log")) {
    log_fs << "t,diam,aspl" << endl;
    log_fs << "0," << best_score.first << ","
           << (double)best_score.second / (n*(n-1)) << endl;
  }
  for(int t = 1; t <= vm["step"].as<int>(); t++) {
    Graph H = G;
    random_swap(&H, conf);
    pair<int, int> curr_score = score(H);
    if(curr_score.first < best_score.first ||
       (curr_score.first == best_score.first &&
        curr_score.second < best_score.second)) {
      G = H;
      best_score = curr_score;
    }
    if(vm.count("log"))
      log_fs << t << "," << best_score.first << ","
             << (double)best_score.second / (n*(n-1)) << endl;
  }

  write_edgelist(G, output);
  if(vm.count("log"))
    log_fs.close();

  return 0;
}
